rllib/models/tf/tf_distributions.py

```diff
@@ -279,13 +290,262 @@ def kl(self, other: "Distribution", **kwargs) -> TensorType:
+                each Categorical distribution. If not provided, 1.0 is used for all.
+            **kwargs: Forward compatibility kwargs.
+        """
+        if not temperatures:
+            # If temperatures are not provided, use 1.0 for all actions.
+            temperatures = [1.0] * len(input_lens)
+
+        categoricals = [
+            TfCategorical(logits=logits, temperature=temperature)
+            for logits, temperature in zip(
+                tf.split(logits, input_lens, axis=1), temperatures
+            )
+        ]
+
+        return TfMultiCategorical(categoricals=categoricals)
+
+    def to_deterministic(self) -> "TfMultiDistribution":
+        return TfMultiDistribution([cat.to_deterministic() for cat in self.cats])
+
+
+@DeveloperAPI
+class TfMultiDistribution(Distribution):
+    """Action distribution that operates on multiple, possibly nested actions."""
+
+    def __init__(
+        self,
+        child_distribution_struct: Union[Tuple, List, Dict],
+    ):
+        """Initializes a TorchMultiActionDistribution object.
+
+        Args:
+            child_distribution_struct: Any struct
+                that contains the child distribution classes to use to
+                instantiate the child distributions from `logits`.
+        """
+        super().__init__()
+        self.original_struct = child_distribution_struct
+        self.flat_child_distributions = tree.flatten(child_distribution_struct)
+
+    @override(Distribution)
+    def rsample(
+        self,
+        *,
+        sample_shape: Tuple[int, ...] = None,
+        **kwargs,
+    ) -> Union[TensorType, Tuple[TensorType, TensorType]]:
+        rsamples = []
+        for dist in self.flat_child_distributions:
+            rsample_logp = dist.rsample(sample_shape=sample_shape, **kwargs)
+            rsample = rsample_logp
+            rsamples.append(rsample)
+
+        rsamples = tree.unflatten_as(self.original_struct, rsamples)
+        return rsamples
+
+    @override(Distribution)
+    def logp(self, value):
+        # Single tensor input (all merged).
+        if isinstance(value, (tf.Tensor, np.ndarray)):
+            split_indices = []
+            for dist in self.flat_child_distributions:
+                if isinstance(dist, TfCategorical):
+                    split_indices.append(1)
+                elif (
+                    isinstance(dist, TfMultiCategorical)
+                    and dist.action_space is not None
+                ):
+                    split_indices.append(np.prod(dist.action_space.shape))
+                else:
+                    sample = dist.sample()
+                    # Cover Box(shape=()) case.
+                    if len(sample.shape) == 1:
+                        split_indices.append(1)
+                    else:
+                        split_indices.append(tf.shape(sample)[1])
+            split_value = tf.split(value, split_indices, axis=1)
```

Comment:
hard coding axis=1 also seems weird. Does this function work for when the input is batched? (B, d) ?